### ISSUE TITLE
[Validator] Inconsistent serialization scheme

### ISSUE THREAD

**Issue by origaminal** [symfony/symfony#15582]
The way how `ClassMetadata` goes to cache seems to be inconsistent. The problem is affected in #12302 

When some code adds to `ClassMetadata` some constraints to **ParentClass** and then **SubClass**'s `ClassMetadata` is built the constraints will appear in that metadata and will have a chance to be cached. Then they could appear in **SubClass**'s `ClassMetadata` in other area of the code where **ParentClass**'s `ClassMetadata` doesn't have the additional runtime constraints.

And vice versa - **ParentClass**'s `ClassMetadata` runtime constraints won't affect **SubClass**'s `ClassMetadata` if it is cached already. Another case is when constraints added to **ParentClass** after **SubClass** is generated won't affect its metadata.

IMO this leads to cases that are hard to debug and hard to bypass. Could it be counted as issue? Are the acceptable ways to resolve the cases?

**Comment by xabbuh** [comment]
@origaminal Can you give a more concrete example that you have in mind?

**Comment by xabbuh** [comment]
ping @origaminal :)

**Comment by origaminal** [comment]
@xabbuh Looks like the described behavior was fixed by https://github.com/symfony/symfony/pull/20793


**Before #20793**. Previously ```ClassMetadata``` was compiled and then it was written to cache. So it could lead to weird behavior. Imagine that cache is clear:
1. You get metadata for class ```Parent``` and add constraints to it;
2. You get metadata for class ```Child```, metadata factory compiles ```ClassMetadata``` of it (e.g. merges its constraint with all including dynamic constraints of ```Parent```) and then writes ```ClassMetadata``` to cache
```php
$parentMetadata = $metadataFactory->getMetadataFor('Parent');
$parentMetadata->addConstraint(new ConstraintA());
$parentMetadata->addConstraint(new ConstraintB());

$childMetadata = $metadataFactory->getMetadataFor('Child');
// ConstraintA and ConstraintB will be in the cached value of childMetadata
```
3. Now if in some other request that is processed by different part of codes ```Child``` metadata will be gotten it will have ``ConstraintA``` and ```ConstraintB``` even if in this part of code such constraints aren't dynamically applied to ```Parent```:
```php
$childMetadata = $metadataFactory->getMetadataFor('Child');
// ConstraintA and ConstraintB are in the $childMetadata constraints while 
// $parentMetadata->addConstraint(new ConstraintA());
// $parentMetadata->addConstraint(new ConstraintB());
// weren't called
```

And vise versa: if the code from item 3 will be called first on empty cache then there will be no ```ConstraintA```, ```ConstraintB``` constraints in ```Child``` metadata in the code from 1-2 items, despite such constraints are dynamically applied to ```Parent``` metadata. Kind of race conditions.


**After #20793** only direct static class constraints are cached. Merging of ancestor constraints into child metadata happens after writing to cache.

**Comment by xabbuh** [comment]
> @xabbuh Looks like the described behavior was fixed by #20793 

Cool 👍 closing here then
